### ISSUE TITLE
broadcast rules not work

### ISSUE THREAD

**Issue by DequanZhu** [NVIDIA/DALI#4758]
I have serialized a dali pipeline and load it by tritonserver, then exec a infer request by tritonclient,but some errors occurs,below is detailed error message
![微信图片_20230331152023](https://user-images.githubusercontent.com/26426479/229052594-f1ff2d9a-2a06-4bc9-84d1-f26e1a0f2e71.png)
the error message hints  me that  I mistakely sub two tensors which shape not exactly  match. According to[ numpy array braodcast rules](https://numpy.org/doc/stable/user/basics.broadcasting.html), 200x2 and 1x2 array are compatible. But on pipeline debug mode, no errors produces, anyone know why?

**Comment by awolant** [comment]
Hello, thanks for reporting the issue.

Could you maybe share the code of your pipeline as well? That would help a lot.

**Comment by DequanZhu** [comment]
Hi, @awolant ,here is a simple example：
for debug mode, data_a is a tensor which shapes 100x2, data_b shapes 1*2 , exec this script , no errors occurs:

```
import numpy as np
import nvidia.dali as dali
import os

numpy_data_a_dir = "/tmp/test_data_a"
numpy_data_b_dir = "/tmp/test_data_b"

data_a = np.random.random(size=(100, 2))
data_b = np.random.random(size=(1, 2))

np.save(os.path.join(numpy_data_a_dir, "data.npy"), data_a)
np.save(os.path.join(numpy_data_b_dir, "data.npy"), data_b)

data_a_files = os.listdir(numpy_data_a_dir)
data_b_files = os.listdir(numpy_data_b_dir)

BATCH_SIZE = 8
@dali.pipeline_def(batch_size=BATCH_SIZE, num_threads=1, device_id=0)
def simple_pipeline():
    test_data_a = dali.fn.readers.numpy(file_root=numpy_data_a_dir, files=data_a_files)
    test_data_b = dali.fn.readers.numpy(file_root=numpy_data_b_dir, files=data_b_files)
    result = test_data_a - test_data_b
    return result

pipe = simple_pipeline(debug=True)
pipe.build()
pipe_out = pipe.run()
print(pipe_out)
```

below is a scipt called dali.py used for serializing it to by loaded by tritonserver, just exec `python3 dali.py  model.dali` :
```
import nvidia.dali as dali

def parse_args():
    import argparse
    parser = argparse.ArgumentParser(description="Serialize the pipeline and save it to a file")
    parser.add_argument('file_path', type=str, help='The path where to save the serialized pipeline')
    return parser.parse_args()


BATCH_SIZE = 8
@dali.pipeline_def(batch_size=BATCH_SIZE, num_threads=1, device_id=0)
def simple_pipeline():
    test_data_a = dali.fn.external_source(device="cpu", name="data_a")
    test_data_b = dali.fn.external_source(device="cpu", name="data_b")
    result = test_data_a - test_data_b
    return result


def main(filename):
    simple_pipeline().serialize(filename=filename)


if __name__ == '__main__':
    args = parse_args()
    main(args.file_path)
```
then load model.dali by tritonserver,the model config is like this:
```
name: "test-broadcast"
backend: "dali"
max_batch_size: 8
input [
  {
    name: "data_a"
    data_type: TYPE_FP32
    dims: [ -1,2 ]
  },
  {
    name: "data_b"
    data_type: TYPE_FP32
    dims: [ -1,2 ]
  }
]

output [
  {
    name: "result"
    data_type: TYPE_FP32
    dims: [-1,2]
  }
]
```
then exec a infer request to model by tritonclient like this:
```
import numpy as np
import tritonclient.http as httpclient
import logging

url = "xxx"
protocol = "http"
model_name = "test-broadcast"

triton_client = httpclient.InferenceServerClient(url=url, verbose=False)

inputs = []
outputs = []
image_output = None
try:
    data_a = np.random.random(size=(1, 100, 2)).astype(np.float32)
    data_b = np.random.random(size=(1, 1, 2)).astype(np.float32)
    inputs.append(httpclient.InferInput('data_a', data_a.shape, "FP32"))
    inputs.append(httpclient.InferInput('data_b', data_b.shape, "FP32"))
    inputs[0].set_data_from_numpy(data_a, binary_data=True)
    inputs[1].set_data_from_numpy(data_b, binary_data=True)
    outputs.append(httpclient.InferRequestedOutput('result', binary_data=True))
    result = triton_client.infer(
        model_name,
        inputs,
        outputs=outputs)
except Exception as e:
    logging.error("infer error", exc_info=True)
else:
    result = result.as_numpy('result')[0]
```
some erros occurs:
![微信图片_20230331164801](https://user-images.githubusercontent.com/26426479/229072819-ecb0d881-5eb8-4429-85cc-aa5593f64a3f.png)

**Comment by mzient** [comment]
It looks like you're using an old version of DALI. Broadcasting is available since DALI 1.20 (Triton dali_backend r23.01).

**Comment by JanuszL** [comment]
@DequanZhu,

Please update the version of the TRITON server you use to at least 23.01, 22.12, and older don't support that yet.
